### ISSUE TITLE
Request: additional options for injecting provider values

### ISSUE THREAD

**Issue by marshall007** [typestack/typedi#23]
Angular DI offers several options for providing dependencies to the container:

```ts
[
  { provide: ..., useClass: FakeLogger },
  { provide: ..., useExisting: Logger},
  { provide: ..., useValue: { log: () => {} } },
  { provide: ..., useFactory: createLogger }
]
```

In particular, it would be great if we could provide a `factory` function. I think this would be much better than the existing solution that requires you use `@Service({ factory: ... })`. With the factory defined in the service decorator it's much more difficult to mock out in tests.

```ts
Container.provide([
  { name: 'foo', value: new Foo() },
  { name: 'foo.class', class: Foo },
  { name: 'foo.factory', factory: () => new Foo() },
])
```

**Comment by pleerock** [comment]
I think right now you can do it via `Container.registerService`, we can add this feature to factory too. This is easy fix and you can do a PR for it.

**Comment by xxxtonixxx** [comment]
I have implemented it in #35 👍

**Comment by pleerock** [comment]
I refactored how `set` method works now you can use it following way:

```ts
            Container.set([
                { id: TestService, value: testService },
                { id: "test1-service", value: test1Service },
                { id: "test2-service", value: test2Service },
                { id: "test3-service", factory: [TestServiceFactory, "create"] },
            ]);
```

My changes were based on @xxxtonixxx changes, thank to him. This will be released in 0.6.0 version.
